File: flutter/amen_client/test/client_test.dart
import 'dart:convert';
import 'package:amen_client/amen_client.dart';
import 'package:http/http.dart' as http;
import 'package:http/testing.dart';
import 'package:test/test.dart';

AmenClient client(Future<http.Response> Function(http.Request) handler) =>
    AmenClient(Config(env: 'sandbox', apiKey: 'test-token', maxRetries: 1), httpClient: MockClient(handler));
http.Response json(int status, Object body) => http.Response(jsonEncode(body), status, headers: {'content-type': 'application/json'});

void main() {
  test('auth header and sandbox base URL', () async {
    final c = client((req) async { expect(req.url.toString(), 'https://sandbox-api.amnn.sa/api/v1/account'); expect(req.headers['X-API-Token'], 'test-token'); return json(200, {'id': 'a1'}); });
    expect(await c.account.get(), {'id': 'a1'});
  });
  test('error codes parsed into AmenApiError', () async {
    final c = client((_) async => json(400, {'error': ['first_name__required']}));
    await expectLater(c.customers.create(firstName: '', lastName: 'x', phoneCode: 'SA', phoneNumber: '5'),
        throwsA(isA<AmenApiError>().having((e) => e.has('first_name__required'), 'has code', true).having((e) => e.retryable, 'retryable', false)));
  });
  test('lifecycle guard blocks invalid action locally', () async {
    var calls = 0;
    final c = client((_) async { calls++; return json(200, {'number': 'DL-1', 'status': 'draft'}); });
    await expectLater(c.deals.actions.approve('DL-1'), throwsA(isA<AmenLifecycleError>()));
    expect(calls, 1);
  });
  test('Origin sent on mutating requests', () async {
    final c = client((req) async { expect(req.headers['Origin'], 'https://sandbox-api.amnn.sa'); return json(201, {'id': 'w', 'url': 'u', 'secret_key': 's'}); });
    expect((await c.webhooks.create('https://example.com/hook')).secretKey, 's');
  });
  test('timestamps are epoch milliseconds', () => expect(toDate(1679568486000)!.year, 2023));
}

File: javascript/tests/unit/client.test.js
import { test } from "node:test";
import assert from "node:assert/strict";
import { AmenClient, AmenApiError, AmenLifecycleError } from "../../src/amen/index.js";
const json = (status, body) => new Response(JSON.stringify(body), { status, headers: { "content-type": "application/json" } });
const client = (f) => new AmenClient({ apiKey: "test-token", baseUrl: "https://sandbox-api.amnn.sa", maxRetries: 1 }, f);

test("auth header + sandbox base URL", async () => {
  const f = async (url, init) => { assert.equal(String(url), "https://sandbox-api.amnn.sa/api/v1/account"); assert.equal(init.headers["X-API-Token"], "test-token"); return json(200, { id: "a1" }); };
  assert.deepEqual(await client(f).account.get(), { id: "a1" });
});
test("error codes parsed into AmenApiError", async () => {
  await assert.rejects(client(async () => json(400, { error: ["first_name__required"] })).customers.create({}), (e) => e instanceof AmenApiError && e.has("first_name__required") && !e.retryable);
});
test("429 retried then succeeds", async () => {
  let calls = 0; const f = async () => (++calls === 1 ? json(429, { error: ["rate_limit__exceeded"] }) : json(200, [{ id: 1 }]));
  const orig = globalThis.setTimeout; globalThis.setTimeout = (cb) => { cb(); return 0; };
  try { assert.deepEqual(await client(f).lookups.cities(), [{ id: 1 }]); assert.equal(calls, 2); } finally { globalThis.setTimeout = orig; }
});
test("lifecycle guard blocks invalid action locally", async () => {
  let calls = 0; const f = async () => { calls++; return json(200, { number: "DL-1", status: "draft" }); };
  await assert.rejects(client(f).deals.actions.approve("DL-1"), AmenLifecycleError); assert.equal(calls, 1);
});
test("Origin sent on mutating requests", async () => {
  const f = async (_u, init) => { assert.equal(init.headers.Origin, "https://sandbox-api.amnn.sa"); return json(201, { id: "w", secret_key: "s" }); };
  assert.equal((await client(f).webhooks.create("https://example.com/hook")).secret_key, "s");
});
